cjworkbench/tests/i18n.py

```diff
@@ -0,0 +1,31 @@
+from django.test import SimpleTestCase
+from cjworkbench.i18n.trans import trans, MessageTranslator, InvalidICUParameters
+from cjworkbench.i18n import default_locale
+
+
+mock_message_id = (
+    "some+crazy+id+that+will+never+be+actually+used+in+real+translation+files"
+)
+
+
+class TransTest(SimpleTestCase):
+    def test_format_invalid_default(self):
+        """Tests that a programmer will get an exception when including a numeric variable in the message
+        """
+        with self.assertRaises(InvalidICUParameters):
+            trans(
+                default_locale,
+                mock_message_id,
+                default="Hello {a} {0} {b}",
+                parameters={"a": "you", "0": "!", "b": "2"},
+            ),
+
+    def test_format_invalid_message(self):
+        """Tests that a translator can't break our system by including a numeric variable in the message
+        """
+        self.assertEqual(
+            MessageTranslator(default_locale).process_message(
+                "Hello {a} {0} {b}", "Hello {a} {b}", parameters={"a": "you", "b": "!"}
+            ),
+            "Hello you !",
+        )
```

Comment:
We need more unit tests:

```
Disallow nesting:
process_message("Allo <a0><a1>x</a1></a0>", "Hello <a0>x</a0><a1>y</a1>", tags={"a0": {"tag": "a": "href": "http://a.com"}, "a1": {"tag": "a", "href": "http://b.com"}})
'Allo <a href="http://a.com">&lt;a1&gt;x&lt;/a1&gt;</a>'
```

```
Substitute parameters within tags:
process_message("Allo <a0>x{y}z</a0>", "Hello <a0>x</a0>{y}", tags={"a0": {"tag": "a", "href": "http://a.com"}}, arguments={"y": "Y"})
'Allo <a href="http://a.com">xYz</a>'
```

```
HTML-escape special characters in message:
process_message("Allo < there", "Hello there")
"Allo &lt; there"
```

```
Avoid injection by translators (really, an explicit _reason_ for escaping HTML from translators):
process_message("Allo <script>x=y</script> there", "Hello there")
"Allo &lt;script&gt;x=y&lt;/script&gt; there"
```

```
HTML-escape parameters:
process_message("Allo {script}", "Hello {script}", parameters={"script": "<script>x=y</script>"})
"Allo &lt;script&gt;x=y&lt;/script&gt;"
```

```
HTML-escape arguments
process_message("Allo <a0>x</a0>", "Hello <a0>x</a0>", tags={"a0": {"tag": "a", "href": "/a?x=y&z=b"}})
'Allo <a href="/a?x=y&amp;z=b">x</a>'
```

... your BeautifulSoup algorithm doesn't return these. What _does_ it return? I feel my proposed return values are sensible and easy to explain to translators: "just write plain text; we'll treat &lt;a0&gt; specially."

Reply:
Most of these properties ('HTML-escape special characters in message', 'Disallow nesting' but see note 1 below, 'Avoid injection by translators') are already tested in the tests for the `trans_html` tag.

1. For nesting, it's easier to just ignore the inner tags (now we keep their contents, but we could not keep them); for example `<a0><span0>Hello</span0></a0>` would become `<a href="something">Hello</a>` (or plain `<a href-"something"></a>`).
2. Arguments and parameters are provided by the programmer, so it's different decision whether to escape them or not.

Reply:
Arguments and parameters are variables. They can be provided by the user. (For instance: a column name might be the argument in a message, and column names are provided by users.)

Reply:
Nope, arguments and parameters are provided _to `trans`_ by the programmer. If the user inputs them somewhere else, isn't it the duty of that "somewhere else" to decide whether to escape them or not?

Reply:
No. It is _never_ sensible to _not_ HTML-escape these arguments. If `trans` didn't HTML-escape all arguments, then programmers would need to HTML-escape arguments every. single. time they call it.

This isn't just theoretical. Your own patch is inconsistent. As a programmer, you coded:

```
{% trans "accounts.sign_up.login_prompt" default='If you have already created an account, then please <a0>log in</a0>.' tag_a0_href=login_url %}
```

`login_url` will often contain an ampersand -- which is invalid HTML. It makes no sense to ask programmers to html_escape every single argument. That requirement would give zero benefit, cost callers space and time, and introduce bugs because programmers would never imagine they need to HTML-escape things.

In the example you wrote, the login URL doesn't contain anything the user writes; but that's just happenstance. Parameters can easily contain user-input data. Imagine:

```
process_message("Download this workflow as a <a0>spreadsheet</a0>", ..., tags={"a0": {"tag": "a", "href": csv_url, "download": filesystem_safe_workflow_title, "title": workflow_title}})
```

In this example, `workflow_title` is provided by Bob, and Alice is viewing the page. I'd expect the programmer to _filesystem-escape_ `filesystem_safe_workflow_title` (removing any `/`). But I wouldn't expect them to HTML-escape it afterwards, because that's the whole point of an HTML-producing library. If Bob names the workflow `r&b`, Alice should see the `download` attribute, `download="r&amp;b.csv`.

And if Bob user names the workflow, `"><script>window.x = y</script>`? I'd expect Alice to see a `title` tag of, `"><script>window.x = y</script>`. If `trans` _doesn't_ HTML-escape, then Bob just injected a script in Alice's browser session.1

`trans` should _never_ output invalid HTML. It should behave as programmers are used to:

```javascript
const a = document.createHTMLElement("a")
a.href = csv_url
a.download = filesystem_safe_workflow_title // no need to HTML-escape
a.title = workflow_title // no need to HTML-escape
document.body.appendChild(a)
```

Reply:
OK, I get it. I updated my pull request; see if it's fine now.

Reply:
Still missing a couple:

* "Disallow nesting"
* "HTML-escape special characters in message"

I'm okay with making the retval from "disallow nesting" something a bit different -- so long as it is utterly impossible to nest tags.

HTML-escaping special characters in the message is essential. Otherwise translators could inject XML entities (which would be one-half of a security vulnerability).

Reply:
These two are tested in `server/tests/templatetags`

Reply:
I moved most tests from `server/tests/templatetags/test_trans.py` to `cjworkbench/tests/i18n.py`. I also added tests for more properties. I think it's fine now.